kolibri/core/device/translation.py

```diff
@@ -3,58 +3,91 @@
 """
 from __future__ import unicode_literals
 
+import re
+
 from django.conf import settings
 from django.core.cache import cache
+from django.urls import resolve
+from django.urls import Resolver404
+from django.urls.resolvers import RegexURLResolver
+from django.utils.translation import get_language
 from django.utils.translation import LANGUAGE_SESSION_KEY
 from django.utils.translation.trans_real import check_for_language
+from django.utils.translation.trans_real import get_language_from_path
 from django.utils.translation.trans_real import get_languages
 from django.utils.translation.trans_real import get_supported_language_variant
 from django.utils.translation.trans_real import language_code_re
 from django.utils.translation.trans_real import parse_accept_lang_header
 
-from .models import DeviceSettings
 
 DEVICE_LANGUAGE_CACHE_KEY = "DEVICE_LANGUAGE_CACHE_KEY"
 
 
+def get_device_language():
+    from .models import DeviceSettings
+
+    try:
+        if cache.get(DEVICE_LANGUAGE_CACHE_KEY) is None:
+            # Use a relatively short expiry, in case the device setting is changed in another
+            # thread and this cache does not get invalidated.
+            cache.set(
+                DEVICE_LANGUAGE_CACHE_KEY, DeviceSettings.objects.get().language_id, 600
+            )
+        return get_supported_language_variant(cache.get(DEVICE_LANGUAGE_CACHE_KEY))
+    except (DeviceSettings.DoesNotExist, LookupError):
+        return None
+
+
+def get_settings_language():
+    try:
+        return get_supported_language_variant(settings.LANGUAGE_CODE)
+    except LookupError:
+        return settings.LANGUAGE_CODE
+
+
 def get_language_from_request(request):  # noqa complexity-16
     """
     Analyzes the request to find what language the user wants the system to
     show. Only languages listed in settings.LANGUAGES are taken into account.
     If the user requests a sublanguage where we have a main language, we send
     out the main language.
-    If check_path is True, the URL path prefix will be checked for a language
-    code, otherwise this is skipped for backwards compatibility.
     """
 
+    try:
+        # If this is not a view that needs to be translated, return None, and be done with it!
+        if not getattr(resolve(request.path_info).func, "translated", False):
+            return None, False
```

Comment:
`The language has been calculated from the path.`

I am using a Russellian theory of descriptions here to return `False`, because of the null referent of `language` (`language` is `None`), the predicate evaluates to `False`, just as `The language has not been calculated from the path` was evaluate to `False` by the same reasoning.

Reply:
my cat's breath smells like cat food